packages/yew-router/src/service.rs

```diff
@@ -1,182 +1,79 @@
-    /// and creates a history entry that can be navigated via the forward and back buttons.
-    ///
-    /// The route should be a relative path that starts with a `/`.
-    pub fn set_route(&mut self, route: &str, state: STATE) {
-        let state_string: String = serde_json::to_string(&state).unwrap_or_else(|_| {
-            log::error!("Could not serialize state string");
-            "".to_string()
-        });
-        let _ = self
-            .history
-            .push_state_with_url(&Value::from_str(&state_string), "", Some(route));
-    }
-
-    /// Replaces the route with another one removing the most recent history event and
-    /// creating another history event in its place.
-    pub fn replace_route(&mut self, route: &str, state: STATE) {
-        let state_string: String = serde_json::to_string(&state).unwrap_or_else(|_| {
-            log::error!("Could not serialize state string");
-            "".to_string()
-        });
-        let _ =
-            self.history
-                .replace_state_with_url(&Value::from_str(&state_string), "", Some(route));
-    }
-
-    /// Gets the concatenated path, query, and fragment.
-    pub fn get_route(&self) -> Route<STATE> {
-        let route_string = Self::get_route_from_location(&self.location);
-        let state: STATE = get_state_string(&self.history)
-            .or_else(|| {
-                log::trace!("History state is empty");
-                None
-            })
-            .and_then(|state_string| -> Option<STATE> {
-                serde_json::from_str(&state_string)
-                    .ok()
-                    .or_else(|| {
-                        log::error!("Could not deserialize state string");
-                        None
-                    })
-                    .and_then(std::convert::identity) // flatten
-            })
-            .unwrap_or_default();
-        Route {
-            route: route_string,
-            state,
-        }
-    }
+    let url = web_sys::Url::new(&yew::utils::document().url().unwrap()).unwrap();
+    let parsed =
+        serde_urlencoded::from_str(&url.search().strip_prefix("?").as_deref().unwrap_or(""))?;
+    Ok(parsed)
 }
 
-/// Formats a path, query, and fragment into a string.
-///
-/// # Note
-/// This expects that all three already have their expected separators (?, #, etc)
-pub(crate) fn format_route_string(path: &str, query: &str, fragment: &str) -> String {
-    format!(
-        "{path}{query}{fragment}",
-        path = path,
-        query = query,
-        fragment = fragment
-    )
+pub fn current_route<R: Routable>() -> Option<R> {
+    let pathname = yew::utils::window().location().pathname().unwrap();
+    R::recognize(&pathname)
 }
 
-fn get_state(history: &History) -> Value {
-    history.state().unwrap()
+// just an opaque handle to the actual listener so we don't expose implementation details
+/// Handle for the router's path event listener
+pub struct RouteListener {
```

Comment:
That comment is no longer necessary.
```suggestion
/// Handle for the router's path event listener
pub struct RouteListener {
```